sql/src/main/java/io/crate/operation/reference/sys/shard/SysAllocations.java

```diff
@@ -0,0 +1,82 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.sys.shard;
+
+import io.crate.operation.reference.sys.CompletableFutureIterableSupplier;
+import org.elasticsearch.action.admin.cluster.allocation.ClusterAllocationExplanation;
+import org.elasticsearch.action.admin.cluster.allocation.TransportClusterAllocationExplainAction;
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.ClusterInfoService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.allocator.ShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.gateway.GatewayAllocator;
+
+import java.util.ArrayList;
+
+@Singleton
+public class SysAllocations extends CompletableFutureIterableSupplier<SysAllocation> {
+
+    private ClusterService clusterService;
+    private ClusterInfoService clusterInfoService;
+    private AllocationDeciders allocationDeciders;
+    private final ShardsAllocator shardAllocator;
+    private final GatewayAllocator gatewayAllocator;
+
+    @Inject
+    public SysAllocations(ClusterService clusterService,
+                          ClusterInfoService clusterInfoService,
+                          AllocationDeciders allocationDeciders,
+                          ShardsAllocator shardAllocator,
+                          GatewayAllocator gatewayAllocator) {
+        this.clusterService = clusterService;
+        this.clusterInfoService = clusterInfoService;
+        this.allocationDeciders = allocationDeciders;
+        this.shardAllocator = shardAllocator;
+        this.gatewayAllocator = gatewayAllocator;
+    }
+
+    @Override
+    protected Iterable<SysAllocation> items() {
+        final ClusterState state = clusterService.state();
+        final RoutingNodes routingNodes = state.getRoutingNodes();
+        final ClusterInfo clusterInfo = clusterInfoService.getClusterInfo();
+
+        final RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, routingNodes, state, clusterInfo,
+            System.nanoTime(), false);
+        ArrayList<SysAllocation> explanations = new ArrayList<>();
```

Comment:
Could we do this lazy instead of creating the list fully up-front?

Reply:
@mfussenegger do you mean by implementing the iterator?

Reply:
yep - an easy way to do that would be `return () -> allocation.routingTable().allShards().stream().map(...).iterator()`